rosbag2_cpp/src/rosbag2_cpp/cache/message_cache.cpp

```diff
@@ -68,18 +71,32 @@ void MessageCache::release_consumer_buffer()
   consumer_buffer_mutex_.unlock();
 }
 
-void MessageCache::swap_buffers()
+void MessageCache::notify_data_ready()
+{
+  {
+    std::lock_guard<std::mutex> lock(producer_buffer_mutex_);
+    data_ready_ = true;
+  }
+  cache_condition_var_.notify_one();
+}
+
+void MessageCache::wait_for_data()
 {
   std::unique_lock<std::mutex> producer_lock(producer_buffer_mutex_);
-  std::lock_guard<std::recursive_mutex> consumer_lock(consumer_buffer_mutex_);
   if (!flushing_) {
     // Required condition check to protect against spurious wakeups
     cache_condition_var_.wait(
       producer_lock, [this] {
-        return primary_buffer_can_be_swapped_ || flushing_;
+        return data_ready_ || flushing_;
```

Comment:
Sanity check (for me): is it possible/likely that `flushing_` will go from `false` to `true` in the time between the check at line 86 to here?